crates/fuzzing/src/generators.rs

```diff
@@ -569,3 +572,92 @@ impl<'a> Arbitrary<'a> for ModuleConfig {
+        match self {
+            CodegenSettings::Native => {}
+            CodegenSettings::Target { target, flags } => {
+                config.target(target).unwrap();
+                for (key, value) in flags {
+                    unsafe {
+                        config.cranelift_flag_set(key, value).unwrap();
+                    }
+                }
+            }
+        }
+    }
+}
+
+impl<'a> Arbitrary<'a> for CodegenSettings {
+    #[allow(unused_macros, unused_variables)]
+    fn arbitrary(u: &mut Unstructured<'a>) -> arbitrary::Result<Self> {
+        // Helper macro to enable clif features based on what the native host
+        // supports. If the input says to enable a feature and the host doesn't
+        // support it then that test case is rejected with a warning.
+        macro_rules! target_features {
+            (
+                test:$test:ident,
+                $(std: $std:tt => clif: $clif:tt $(range: $e:expr)?,)*
+            ) => ({
+                let mut flags = Vec::new();
+                $(
+                    let range = 0..=1;
+                    $(let range = { drop(range); $e };)?
+                    let enable = u.int_in_range(range)? == 1;
+                    if enable && !std::$test!($std) {
+                        log::error!("want to enable clif `{}` but host doesn't support it",
+                            $clif);
+                        return Err(arbitrary::Error::EmptyChoose)
+                    }
+                    flags.push((
+                        $clif.to_string(),
+                        enable.to_string(),
+                    ));
+                )*
+                flags
+            })
+        }
+        #[cfg(target_arch = "x86_64")]
+        {
+            if u.int_in_range(0..=9)? == 0 {
+                let flags = target_features! {
+                    test: is_x86_feature_detected,
+                    std:"sse3" => clif:"has_sse3",
+                    std:"ssse3" => clif:"has_ssse3",
+                    std:"sse4.1" => clif:"has_sse41",
+                    std:"sse4.2" => clif:"has_sse42",
+                    std:"popcnt" => clif:"has_popcnt",
+                    std:"avx" => clif:"has_avx",
+                    std:"avx2" => clif:"has_avx2",
+                    std:"bmi1" => clif:"has_bmi1",
+                    std:"bmi2" => clif:"has_bmi2",
+                    std:"lzcnt" => clif:"has_lzcnt",
+
+                    // not a lot of of cpus support avx512 so these are weighted
+                    // to get enabled much less frequently.
+                    std:"avx512bitalg" => clif:"has_avx512bitalg" range:0..=1000,
+                    std:"avx512dq" => clif:"has_avx512dq" range: 0..=1000,
+                    std:"avx512f" => clif:"has_avx512f" range: 0..=1000,
+                    std:"avx512vl" => clif:"has_avx512vl" range: 0..=1000,
+                    std:"avx512vbmi" => clif:"has_avx512vbmi" range: 0..=1000,
```

Comment:
By the way, I added a method upstream the other day that can make this a little nicer: https://docs.rs/arbitrary/latest/arbitrary/struct.Unstructured.html#method.ratio

(fwiw, I also added this method too: https://docs.rs/arbitrary/latest/arbitrary/struct.Unstructured.html#method.arbitrary_loop)

Reply:
Nice!